src/main/java/org/broadinstitute/hellbender/tools/spark/transforms/markduplicates/MarkDuplicatesSparkUtils.java

```diff
@@ -25,76 +32,154 @@
+
+            // If there are two primary reads in the group pass them as a pair
+            } else if (primaryReads.size()==2) {
+                final IndexPair<GATKRead> firstRead = primaryReads.get(0);
+                final IndexPair<GATKRead> secondRead = primaryReads.get(1);
+                final PairedEnds pair = MarkDuplicatesSparkRecord.newPair(firstRead.getValue(), secondRead.getValue(), header, secondRead.getIndex(), scoringStrategy);
                 out.add(new Tuple2<>(pair.key(header), pair));
+
+            // If there is one paired read in the template this probably means the bam is missing its mate, don't duplicate mark it
+            } else if (primaryReads.size()==1) {
+                final IndexPair<GATKRead> firstRead = primaryReads.get(0);
+                final MarkDuplicatesSparkRecord pass = MarkDuplicatesSparkRecord.getPassthrough(firstRead.getValue(), firstRead.getIndex());
+                out.add(new Tuple2<>(pass.key(header), pass));
+
+            // else that means there are no non-secondary or supplementary reads, thus we want the group to pass through through unmarked
+            } else {
+                if (hadNonPrimaryRead.get() !=null) {
+                    final MarkDuplicatesSparkRecord pass = MarkDuplicatesSparkRecord.getPassthrough(hadNonPrimaryRead.get().getValue(), hadNonPrimaryRead.get().getIndex());
+                    out.add(new Tuple2<>(pass.key(header), pass));
+                }
             }
+
             return out.iterator();
-        }).groupByKey(numReducers);
+        });
+
+        final JavaPairRDD<Integer, Iterable<MarkDuplicatesSparkRecord>> keyedPairs = pairedEnds.groupByKey(); //TODO evaluate replacing this with a smart aggregate by key.
 
-        return markPairedEnds(keyedPairs, scoringStrategy, finder, header);
+        return markDuplicateRecords(keyedPairs, finder);
     }
 
-    static JavaPairRDD<String, Iterable<GATKRead>> spanReadsByKey(final SAMFileHeader header, final JavaRDD<GATKRead> reads) {
-        JavaPairRDD<String, GATKRead> nameReadPairs = reads.mapToPair(read -> new Tuple2<>(read.getName(), read));
+    //todo use this instead of keeping all unmapped reads as non-duplicate
+    public static boolean readAndMateAreUnmapped(GATKRead read) {
+        return read.isUnmapped() && (!read.isPaired() || read.mateIsUnmapped());
+    }
+
+    private static JavaPairRDD<String, Iterable<IndexPair<GATKRead>>> getReadsGroupedByName(SAMFileHeader header, JavaRDD<GATKRead> reads, int numReducers) {
+
+        final JavaPairRDD<String, Iterable<IndexPair<GATKRead>>> keyedReads;
+        final JavaRDD<IndexPair<GATKRead>> indexedReads = reads.mapPartitionsWithIndex(
+                (index, iter) -> Utils.stream(iter).map(read -> new IndexPair<>(read, index)).iterator(), false);
+        if (SAMFileHeader.SortOrder.queryname.equals(header.getSortOrder()) ||  SAMFileHeader.GroupOrder.query.equals(header.getGroupOrder()) ) {
```

Comment:
This boolean expression (`SAMFileHeader.SortOrder.queryname.equals(header.getSortOrder()) ||  SAMFileHeader.GroupOrder.query.equals(header.getGroupOrder()`) is used verbatim in several places and should probably be pulled out into a one-line static predicate method.